File: helloFriend/helloFriends.py
def greet_close_friends(friends):
    """
    If you greet both male and female friends in the same manner,
    then this is fine.
    """
    for friend in friends:
        print("I hug {}\n".format(friend))

def greet_not_so_close_friends(friends):
    for friend in friends:
        print("Hey {}?! How have you been?\n".format(friend))

def greet_strangers(strangers):
    for stranger in strangers:
        print("Hi {}. My name is Marcus\n".format(stranger))

def greet_males(males):
    for male in males:
        print("Give {} a firm handshake while looking him in the eyes".format(male))

def greet_females(females):
    for female in females:
        print("Give {} a slight wave while telling her my name".format(female))

if __name__ == "__main__":
    close_friends = ["David", "Christy", "Carmela", "Evelyn"]
    not_so_close_friends = ["Tony", "Ivan", "Alex", "Julie"]
    strangers = ["UnknownGuy", "UnknownGirl"]
    males = ["David", "Tony", "Alex", "Ivan", "UnknownGuy"]
    females = ["Christy", "Carmela", "Evelyn", "Julie", "UnknownGirl"]
    
    print("{}:".format(greet_close_friends.__name__))
    greet_close_friends(close_friends)

    print("\n{}:".format(greet_not_so_close_friends.__name__))
    greet_not_so_close_friends(not_so_close_friends)

    print("\n{}:".format(greet_strangers.__name__))
    greet_strangers(strangers)

    print("\n{}:".format(greet_males.__name__))
    greet_males(males)

    print("\n{}:".format(greet_females.__name__))
    greet_females(females)

File: helloFriend/helloFriends.go
package main

import (
	"fmt"
)

func printToScreen(stringSlice []string) {
	for _, _string := range stringSlice {
		fmt.Println(_string)
	}
	fmt.Printf("\n\n")
}

func greetCloseFriend(friend string) string {
	return fmt.Sprintf("I hug %s", friend)
}

func greetNotSoCloseFriend(friend string) string {
	return fmt.Sprintf("Hey %s?! How have you been?", friend)
}

func greetStranger(stranger string) string {
	return fmt.Sprintf("Hi %s. My name is Marcus", stranger)
}

func greetMale(male string) string {
	return fmt.Sprintf("Give %s a firm handshake while looking him in the eyes", male)
}

func greetFemale(female string) string {
	return fmt.Sprintf("Give %s a slight wave while telling her my name", female)
}

// I need to rethink how I am going to write this...
func greetAPerson(name string, people map[string][]string, genders map[string][]string) {
	var typeOfFriend string // relationship of person
	var gender string  // gender of person
	var greetings []string //

    // Checks to see if there is a typeOfFriend relationship
    // list for the name of the parson that was passed in.
	for key, values := range people {
		for _, value := range values {
			if name == value {
				typeOfFriend = key
				break
			}
		}
        // This is used to break out of the outer for loop.
		if typeOfFriend != "" {
			break
		}
	}

	// If a relationship was not found, make them a stranger.
	if typeOfFriend == "" {
		typeOfFriend = "stranger"
	}

    // Checks to see if we know the gender for this person.
	for key, values := range genders {
		for _, value := range values {
			if name == value {
				gender = key
				break
			}
		}
        // Used to break out of the outer for loop.
		if gender != "" {
			break
		}
	}

	// add if; else if; else statement for calling the right function
	if typeOfFriend == "closeFriend" {
		greetings = append(greetings, greetCloseFriend(name))
	} else if typeOfFriend == "notSoCloseFriend" {
		greetings = append(greetings, greetNotSoCloseFriend(name))
	} else if typeOfFriend == "stranger" {
		greetings = append(greetings, greetStranger(name))

		if gender == "male" {
			greetings = append(greetings, greetMale(name))
		} else if gender == "female" {
			greetings = append(greetings, greetFemale(name))
		}
	}
	printToScreen(greetings)
}

func main() {
	closeFriends := []string{"David", "Christy", "Carmela", "Evelyn"}
	notSoCloseFriends := []string{"Tony", "Ivan", "Alex", "Julie"}
	strangers := []string{"UnknownGuy", "UnknownGirl"}
	males := []string{"David", "Tony", "Alex", "Ivan", "UnknownGuy"}
	females := []string{"Christy", "Carmela", "Evelyn", "Julie", "UnknownGirl"}

	people := map[string][]string{
		"closeFriend":      closeFriends,
		"notSoCloseFriend": notSoCloseFriends,
		"stranger":         strangers,
	}

	genders := map[string][]string{
		"male":   males,
		"female": females,
	}

	greetAPerson("David", people, genders)
	greetAPerson("UnknownGirl", people, genders)
}
